### PR TITLE
[SIL] Witness methods store the conformance from which they come.

### PR DESCRIPTION
**Pull request by huonw**

### CODE REVIEW

File: lib/SIL/SILFunctionType.cpp [hunk 1]
```diff
@@ -2213,12 +2225,26 @@ namespace {
...
 
-      return SILFunctionType::get(nullptr,
-                                  origType->getExtInfo(),
-                                  origType->getCalleeConvention(),
-                                  substParams, substResults,
-                                  substErrorResult,
-                                  getASTContext());
+      Optional<ProtocolConformanceRef> witnessMethodConformance;
+      if (auto conformance = origType->getWitnessMethodConformanceOrNone()) {
+        if (conformance->isAbstract()) {
```
Comment (huonw) [inline lib/SIL/SILFunctionType.cpp]: @slavapestov I don't understand what substitution I'm meant to being here to convert the abstract conformance into a concrete one (and even for the concrete case, I'm just guessing).

**Comment by huonw** [comment]
@swift-ci please smoke test

**Comment by huonw** [comment]
@swift-ci please smoke test

**Comment by huonw** [comment]
@swift-ci please smoke test

**Comment by huonw** [comment]
@swift-ci please smoke test.

**Comment by huonw** [comment]
@swift-ci please smoke test.

**Comment by huonw** [comment]
@swift-ci please smoke test.

File: lib/SIL/SILFunctionType.cpp [hunk 2]
```diff
@@ -2213,12 +2225,27 @@ namespace {
...
-                                  origType->getExtInfo(),
-                                  origType->getCalleeConvention(),
-                                  substParams, substResults,
-                                  substErrorResult,
-                                  getASTContext());
+      Optional<ProtocolConformanceRef> witnessMethodConformance;
+      if (auto conformance = origType->getWitnessMethodConformanceOrNone()) {
+        assert(origType->getExtInfo().hasSelfParam());
+        auto selfType = origType->getSelfParameter().getType();
+        // The Self type can be nested in a few layers of metatypes (etc.), e.g.
```
Comment (huonw) [inline lib/SIL/SILFunctionType.cpp]: Example:

```swift
protocol Foo {
  static var x: Int { get set }
}

func mutate(_: inout Int) {}
func foo<T: Foo>(ty: T.Type) {
  mutate(&T.x)
}

```

The SIL for the setter callback call after calling materializeForSet looks like:

```
...
bb2(%16 : $Builtin.RawPointer):                   // Preds: bb0
  %17 = pointer_to_thin_function %16 : $Builtin.RawPointer to $@convention(witness_method: Foo) <τ_0_0 where τ_0_0 : Foo> (Builtin.RawPointer, @inout Builtin.UnsafeValueBuffer, @inout @thick τ_0_0.Type, @thick τ_0_0.Type.Type) -> () // user: %22
...
```

**Comment by huonw** [comment]
@swift-ci please smoke test and merge.

**Comment by huonw** [comment]
@swift-ci please smoke test and merge.

**Comment by huonw** [comment]
@swift-ci please smoke test and merge.

**Comment by huonw** [comment]
@swift-ci please smoke test and merge.

**Comment by huonw** [comment]
@swift-ci please smoke test and merge.

**Comment by slavapestov** [comment]
Are you planning on doing serialization in a separate commit? If so LGTM!

File: include/swift/AST/Types.h [hunk 3]
```diff
@@ -3547,6 +3550,20 @@ class SILFunctionType final : public TypeBase, public llvm::FoldingSetNode,
...
+  /// for which the method is a witness.
+  ProtocolConformanceRef getWitnessMethodConformance() const {
+    assert(getRepresentation() == Representation::WitnessMethod);
+    return *WitnessMethodConformance;
+  }
+
+  /// If this is a @convention(witness_method) function, return the conformance
+  /// for which the method is a witness, if it isn't that convention, return
+  /// None.
+  Optional<ProtocolConformanceRef> getWitnessMethodConformanceOrNone() const {
```
Comment (slavapestov) [inline include/swift/AST/Types.h]: Is this variant useful?

**Comment by huonw** [comment]
Serialization is in the first commit.

File: include/swift/AST/Types.h (hunk 3 above)
Comment (huonw) [reply include/swift/AST/Types.h]: It's the one I call most often: it avoids having do a `fnTy->getRepresentation() == SILFunctionRepresentation::WitnessMethod ? fnTy.getWitnessMethodConformance() : None` dance.

**Review by slavapestov — APPROVED** [review]

File: include/swift/AST/Types.h (hunk 3 above)
Comment (slavapestov) [reply include/swift/AST/Types.h]: I see, OK.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by huonw** [comment]
@swift-ci Please smoke test

**Comment by huonw** [comment]
@swift-ci Please smoke test

**Comment by huonw** [comment]
@swift-ci Please smoke test.

**Comment by huonw** [comment]
Failure on linux is:

```
Stack dump:
0.	Program arguments: /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swift-linux-x86_64/bin/swift -frontend -c /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/swiftpm/Sources/PackageGraph/DependencyResolver.swift /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/swiftpm/Sources/PackageGraph/PackageGraph.swift /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/swiftpm/Sources/PackageGraph/PackageGraphLoader.swift /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/swiftpm/Sources/PackageGraph/PackageGraphRoot.swift /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/swiftpm/Sources/PackageGraph/RawPackageConstraints.swift /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/swiftpm/Sources/PackageGraph/RepositoryPackageContainerProvider.swift -target x86_64-unknown-linux -disable-objc-interop -sdk / -I /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release -I /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/swiftpm/Sources/clibc/include -I /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/foundation-linux-x86_64/Foundation -I /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/foundation-linux-x86_64/Foundation/usr/lib/swift -I /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/xctest-linux-x86_64 -I /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/libdispatch-linux-x86_64/src -I /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/libdispatch-linux-x86_64/src/swift -I /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/swift-corelibs-libdispatch -enable-testing -module-cache-path /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/ModuleCache -swift-version 4 -O -D SWIFT_PACKAGE -Xcc -fmodule-map-file=/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/swiftpm/Sources/clibc/include/module.modulemap -Xcc -fblocks -emit-module-doc-path /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/PackageGraph.swiftdoc -parse-as-library -module-name PackageGraph -emit-module-path /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/PackageGraph.swiftmodule -emit-dependencies-path /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/PackageGraph.build/PackageGraph.d -num-threads 32 -o /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/PackageGraph.build/DependencyResolver.swift.o -o /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/PackageGraph.build/PackageGraph.swift.o -o /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/PackageGraph.build/PackageGraphLoader.swift.o -o /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/PackageGraph.build/PackageGraphRoot.swift.o -o /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/PackageGraph.build/RawPackageConstraints.swift.o -o /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/PackageGraph.build/RepositoryPackageContainerProvider.swift.o 
1.	While running pass #61987 SILFunctionTransform "Indirect Call Devirtualization" on SILFunction "@_T0s8SequencePss9Equatable7ElementRpzrlE8containsSbAEFSay12PackageGraph22ResolvedProductBuilder33_04246CEF6CEAB47798DA707D73AF3E55LLCG_Tg5".
 for 'contains(_:)' in module 'Swift'
<unknown>:0: error: unable to execute command: Aborted
```

**Comment by huonw** [comment]
@swift-ci Please smoke test.

**Comment by huonw** [comment]
@swift-ci Please smoke test.

**Comment by huonw** [comment]
@swift-ci smoke test

**Comment by huonw** [comment]
@swift-ci smoke test

**Comment by huonw** [comment]
@swift-ci smoke test.

Maybe this will work some day?!

**Comment by shahmishal** [comment]
@swift-ci smoke test

**Comment by huonw** [comment]
@swift-ci smoke test.

**Comment by huonw** [comment]
@swift-ci smoke test

**Comment by huonw** [comment]
@swift-ci smoke test

**Comment by huonw** [comment]
@swift-ci smoke test

**Comment by huonw** [comment]
@swift-ci smoke test

**Comment by huonw** [comment]
@swift-ci smoke test and merge

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by shahmishal** [comment]
@swift-ci Please smoke test Linux

**Comment by shahmishal** [comment]
@swift-ci Please smoke test Linux

**Comment by huonw** [comment]
@swift-ci smoke test and merge

**Comment by huonw** [comment]
@swift-ci smoke test and merge

**Comment by huonw** [comment]
@swift-ci Please smoke test and merge

**Comment by huonw** [comment]
@swift-ci Please smoke test

**Comment by huonw** [comment]
@swift-ci Please smoke test

**Comment by huonw** [comment]
@swift-ci Please smoke test

**Pull request merged**